### DISCUSSION TITLE
Trouble with children in custom "admonition" plugin in compiler

### DISCUSSION THREAD

**Discussion by Firgrep** [mdx-js/mdx#2355] [Q&A] [answered]
Hi there,

I'm having some issues with passing children in a plugin made for directives. I'm using [MDX Editor](https://mdxeditor.dev/) to write the mdx, which, upon hitting a save button, sends the mdx string to the compiler.

I've almost got it working but in order for the admonition to appear as it does in the MDX Editor, I need to insert an additional `<div>` between the directive `<div>` and the contents. I want to keep the contents as they are since there may be a lot of different things coming in, and so I try to just pass the node.children as the children of this new child.

So basically, all I want to do is to insert a new `<div>`  between the directive's `<div>` and its children.

But when I pass the `node.children` into the `h()` function, the compilation doesn't complete (I get an error thrown in the function that calls the compiler). However, if I pass a string as children into the `h()`, then everything works fine.

Here is the compiler, the plugin, and the mdx source.

```ts
import {compile} from '@mdx-js/mdx'
import remarkGfm from "remark-gfm";
import remarkDirective from 'remark-directive';

export const mdxCompiler = async (mdxSource: string) => {
    try {
        const mdxCompiled = String(await compile(mdxSource, {
            outputFormat: 'function-body',
            development: false,
            remarkPlugins: [remarkGfm, remarkDirective, adminitionPlugin],
        }));
        console.log("COMPILED: \n", mdxCompiled) // this line does not get hit
        return mdxCompiled;
    } catch(e) {
        throw new Error("An error occured in the mdxCompiler");
    }
}

import {h} from 'hastscript'
import {visit} from 'unist-util-visit'

const ADMONITION_TYPES = ["note", "tip", "danger", "info", "caution"]
function adminitionPlugin() {
    // @ts-ignore
    return (tree) => {
        visit(tree, (node) => {
            if (
                node.type === 'containerDirective' ||
                node.type === 'leafDirective' ||
                node.type === 'textDirective'
            ) {
                if (ADMONITION_TYPES.includes(node.name)) {

                    const data = node.data || (node.data = {});
                    const tagName = node.type === 'textDirective' ? 'span' : 'div';
                    
                    if (node.name === ADMONITION_TYPES[0]) {
                        node.attributes = {
                            ...node.attributes,
                            class: '_admonitionNote_13nbk_63', 
                        };
                    } else if (node.name === ADMONITION_TYPES[1]) {
                        node.attributes = {
                            ...node.attributes,
                            class: '_admonitionTip_13nbk_63', 
                        };
                    } else if (node.name === ADMONITION_TYPES[2]) {
                        node.attributes = {
                            ...node.attributes,
                            class: '_admonitionDanger_13nbk_63', 
                        };
                    } else if (node.name === ADMONITION_TYPES[3]) {
                        node.attributes = {
                            ...node.attributes,
                            class: '_admonitionInfo_13nbk_63', 
                        };
                    } else if (node.name === ADMONITION_TYPES[4]) {
                        node.attributes = {
                            ...node.attributes,
                            class: '_admonitionCaution_13nbk_63', 
                        };
                    }
                    
                    data.hName = tagName;
                    data.hProperties = h(tagName, node.attributes || {}).properties;
                    console.log("NODE CHILDREN: \n", node.children)
                    // inserting new child to node.data and passing node.data's children to this new child
                    data.hChildren = [
                        h('div', {class: "_nestedEditor_w1wlt_891"}, node.children) 
                                // if i swap node.children here with a string, the compilation runs fine
                    ];
                    console.log("DATA: \n", data)
                    console.log("old children under new child of node.data: \n", data.hChildren[0].children)
                }
            }
        })
    }
}
```

The mdx source: 
```md
# Supercool big title

:::danger
danger multi-line admonition

here is the second line

and third
:::

last sentence
```

Logs from the console:
```ts
// this is prior the h() function being called, and so these are the children of the directive
NODE CHILDREN:
 [
  {
    type: 'paragraph',
    children: [ [Object] ],
    position: { start: [Object], end: [Object] }
  },
  {
    type: 'paragraph',
    children: [ [Object] ],
    position: { start: [Object], end: [Object] }
  },
  {
    type: 'paragraph',
    children: [ [Object] ],
    position: { start: [Object], end: [Object] }
  }
]

// this is after the h() function is called, so the directive gets a new class attribute as well as
// a new child div, which itself is supposed to take over its children
DATA:
 {
  hName: 'div',
  hProperties: { className: [ '_admonitionDanger_13nbk_63' ] },
  hChildren: [
    {
      type: 'element',
      tagName: 'div',
      properties: [Object],
      children: [Array]
    }
  ]
}

// inspecting the transfered children, it appears just as it was in the first console log
// which is expected
old children under new child of node.data:
 [
  {
    type: 'paragraph',
    children: [ [Object] ],
    position: { start: [Object], end: [Object] }
  },
  {
    type: 'paragraph',
    children: [ [Object] ],
    position: { start: [Object], end: [Object] }
  },
  {
    type: 'paragraph',
    children: [ [Object] ],
    position: { start: [Object], end: [Object] }
  }
]
```

So I'm really not sure what's going here. Thanks for any help!

**Comment by ChristianMurphy** [comment]
Welcome @Firgrep! 👋 
Sorry you ran into a spot of trouble.
This looks very similar to https://github.com/orgs/remarkjs/discussions/1218
The issue you are facing is the `hChildren` prop accepts an HTML AST (`hast`), but you are passing a Markdown AST (`mdast`). https://github.com/syntax-tree/mdast-util-to-hast#hchildren

The easier approach to solve this would be to use plain `children` prop instead which does accept `mdast`.

**Reply by Firgrep** [reply]
Hey @ChristianMurphy ! Thanks so much for the welcome and for helping out!

Alright, I'm getting a better sense of what's going on. I tried to switch out `hChildren` with `children`, but now I get the children but not the inserted `<div>`. Any clues?

```ts
                    data.children = [
                        h('div', {class: "_nestedEditor_w1wlt_891"}, node.children)
                    ];
```

I then tried:
```ts
                    data.children = [
                        {
                            type: 'element',
                            tagName: 'div',
                            properties: {className: ["_nestedEditor_w1wlt_891"]},
                            children: node.children
                        }
                    ];
```

And then this:
```ts
                    data.children = [
                        {
                            hName: 'div',
                            hProperties: { className: [ '_nestedEditor_w1wlt_891' ] },
                            children: node.children
                        }
                    ];
```

**Reply by ChristianMurphy** [reply]
When using `children` the inserted `div` needs to start as a Markdown element/`mdast`.
https://github.com/syntax-tree/mdast#readme

Your last example is the closest, but needs a `type` attribute to be valid `mdast`.
```js
data.children = [
    {
        type: 'paragraph',
        hName: 'div',
        hProperties: { className: [ '_nestedEditor_w1wlt_891' ] },
        children: node.children
    }
];
```

**Reply by Firgrep** [reply]
Thanks for coming back quickly on this. Ah, I see, I first understood that `paragraph` to mean "inject HTML `<p>` here". 

Okay, so I've added the `type: 'paragraph'` and... same result. I then tried `type: 'html'` and that didn't change anything either.

**Reply by ChristianMurphy** [reply]
Could you share a runnable example of what you are seeing?
Here are some sandbox starters which could help:

| description | codesandbox | stackblitz |
| - | - | - |
| MDX Loader with Next.js | [codesandbox](https://codesandbox.io/s/github/mdx-js/.github/tree/main/sandbox-templates/mdx-loader-next) | [stackblitz](https://stackblitz.com/github/mdx-js/.github/tree/main/sandbox-templates/mdx-loader-next) | 
| MDX Evaluate with Node.js | [codesandbox](https://codesandbox.io/s/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaulate-node) | [stackblitz](https://stackblitz.com/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaulate-node) |
| MDX Evaluate with Create React App | [codesandbox](https://codesandbox.io/s/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaluate-create-react-app) | [stackblitz](https://stackblitz.com/github/mdx-js/.github/tree/main/sandbox-templates/mdx-evaluate-create-react-app) |

**Reply by Firgrep** [reply]
I couldn't get the Node.js box to work, so I try with create-react-app, and even though I'm getting React errors, you can see the compiler's output in the browser console. It appears the same as on my local app: the new div is missing. 

https://stackblitz.com/edit/github-qo53wd?file=src%2FApp.js

**Reply by ChristianMurphy** [reply]
@Firgrep thanks for sharing, is not with MDX, it is that React does not support `async` hooks.
Some commentary on that here: https://ultimatecourses.com/blog/using-async-await-inside-react-use-effect-hook

A working example would be: https://stackblitz.com/edit/github-qo53wd-ati5ag?file=src%2FApp.js

<details>
<summary>source from example</summary>

```js

import { useEffect, useState } from 'react';
import * as runtime from 'react/jsx-runtime';
import { evaluate } from '@mdx-js/mdx';
import remarkGfm from 'remark-gfm';
import remarkDirective from 'remark-directive';
import { visit } from 'unist-util-visit';
const ADMONITION_TYPES = ['note', 'tip', 'danger', 'info', 'caution'];

function adminitionPlugin() {
  return (tree) => {
    visit(tree, (node) => {
      if (
        node.type === 'containerDirective' ||
        node.type === 'leafDirective' ||
        node.type === 'textDirective'
      ) {
        if (ADMONITION_TYPES.includes(node.name)) {
          const data = node.data || (node.data = {});
          const tagName = node.type === 'textDirective' ? 'span' : 'div';

          if (node.name === ADMONITION_TYPES[0]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionNote_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[1]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionTip_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[2]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionDanger_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[3]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionInfo_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[4]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionCaution_13nbk_63',
            };
          }

          data.hName = tagName;
          data.hProperties = node.attributes;
          console.log('NODE CHILDREN: \n', node.children);

          data.children = [
            {
              type: 'paragraph',
              hName: 'div',
              hProperties: { className: ['_nestedEditor_w1wlt_891'] },
              children: node.children,
            },
          ];

          console.log('DATA: \n', data);
          console.log(
            'old children under new child of node.data: \n',
            node.children
          );
        }
      }
    });
  };
}

const mdxContent = `
# Supercool big title

:::danger
danger multi-line admonition

here is the second line

and third
:::

last sentence
`;

export function useMDX(mdxSource) {
  const [component, setComponent] = useState('div');
  useEffect(() => {
    const render = async () => {
      try {
        const { default: Content } = await evaluate(mdxSource, {
          ...runtime,
          ...{ remarkPlugins: [remarkGfm, remarkDirective, adminitionPlugin] },
          development: false,
        });
        console.log('component', Content);
        setComponent(Content);
      } catch (e) {
        throw new Error('An error occured in the mdxCompiler');
      }
    };

    render();
  }, [mdxSource]);
  console.log('component', component);
  return component;
}

export default function App() {
  const Component = useMDX(mdxContent);
  return Component;
}
```


</details>

**Reply by Firgrep** [reply]
@ChristianMurphy thanks for cleaning that up and making a working example. I'll have a look at the commentary about async and React!

Have you had any further ideas about the div insertion? I notice that in the working sample, if I add an empty array to `node.children`, the children still appear to render. If change it to, `node.hChildren` they do not render (as expected from an empty array).

**Reply by ChristianMurphy** [reply, marked as answer]
Focus on the distinction between what properties can be applied directly to `node` and which go on `node.data`
`children` is a property directly of node.
`hProperties` and `hName` are of `node.data`.
ref: https://github.com/syntax-tree/mdast#readme and https://github.com/syntax-tree/mdast-util-to-hast#fields-on-nodes

Those were flipped in a few spots causing bugs.
I'd highly recommend using TypeScript or JSDoc (with TypeScript validation) to help catch mismatch in object structures.
With the data put in the correct spot, it adds the elements as expected:

https://stackblitz.com/edit/github-qo53wd-7vuhzr?file=src%2FApp.js

<details>
<summary>source from sandbox</summary>

```js
import { useEffect, useState } from 'react';
import * as runtime from 'react/jsx-runtime';
import { evaluate } from '@mdx-js/mdx';
import remarkGfm from 'remark-gfm';
import remarkDirective from 'remark-directive';
import { visit } from 'unist-util-visit';
const ADMONITION_TYPES = ['note', 'tip', 'danger', 'info', 'caution'];

function adminitionPlugin() {
  return (tree) => {
    visit(tree, (node) => {
      if (
        node.type === 'containerDirective' ||
        node.type === 'leafDirective' ||
        node.type === 'textDirective'
      ) {
        if (ADMONITION_TYPES.includes(node.name)) {
          node.data = node.data ?? {};
          const tagName = node.type === 'textDirective' ? 'span' : 'div';

          if (node.name === ADMONITION_TYPES[0]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionNote_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[1]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionTip_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[2]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionDanger_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[3]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionInfo_13nbk_63',
            };
          } else if (node.name === ADMONITION_TYPES[4]) {
            node.attributes = {
              ...node.attributes,
              className: '_admonitionCaution_13nbk_63',
            };
          }

          node.type = 'paragraph';
          node.data = {
            ...node.data,
            hName: tagName,
            hProperties: node.attributes,
          };

          node.children = [
            {
              type: 'paragraph',
              data: {
                hName: 'div',
                hProperties: { className: ['_nestedEditor_w1wlt_891'] },
              },
              children: node.children,
            },
          ];
        }
      }
    });
  };
}

const mdxContent = `
# Supercool big title

:::danger
danger multi-line admonition

here is the second line

and third
:::

last sentence
`;

export function useMDX(mdxSource) {
  const [component, setComponent] = useState('div');
  useEffect(() => {
    const render = async () => {
      try {
        const { default: Content } = await evaluate(mdxSource, {
          ...runtime,
          ...{ remarkPlugins: [remarkGfm, remarkDirective, adminitionPlugin] },
          development: false,
        });
        setComponent(Content);
      } catch (e) {
        throw new Error('An error occured in the mdxCompiler');
      }
    };

    render();
  }, [mdxSource]);
  return component;
}

export default function App() {
  const Component = useMDX(mdxContent);
  return Component;
}

```

</details>

**Reply by Firgrep** [reply]
@ChristianMurphy OK, I've kept tinkering with it and I think I've maybe got something. I had a look at the mdx playground and looked at what the mdast produced when I put a `<div></div>` into the mdx. I also removed any hStuff just in case that was confusing the compiler, and after that `node.children` reacted as expected. I've afterwards re-instated the hastscript stuff related to the top-parent div in order to change the attributes. 

I want to check and see if I can get this working on my local app, but this is promising!

https://stackblitz.com/edit/github-qo53wd-nvcbw5?file=src%2FApp.js

<details>
<summary>Main change/addition</summary>
```ts
node.children = [
            {
              type: "mdxJsxFlowElement",
              name: "div",
              attributes: [
                {
                  type: "mdxJsxAttribute",
                  name: "class",
                  value: "inner_admonition_class_thing"
                }
              ],
              children: node.children,
            }
          ]
```
</details>

**Reply by Firgrep** [reply]
@ChristianMurphy Sorry, missed your latest reply! Ah, I see how I've not really had a clear grasp of node and node.data. I'm looking at your example now and it's very helpful. I think that should sort my problem!

Yes, I do use TypeScript on my local app, but I had its typechecking switched off for the admonition function after it was complaining that the type i added for `tree` was excessively large or possibly infinite. I did find the types for `node` in the first stackblitz you referenced above and I had those put in yesterday. 

Thanks so much @ChristianMurphy for your time and patience. 🙏 I've learned a lot more about how mdx works and I'm excited to about the possibilities it offers.

**Reply by ChristianMurphy** [reply]
> Yes, I do use TypeScript on my local app, but I had its typechecking switched off for the admonition function after it was complaining that the type i added for tree was excessively large or possibly infinite. 

🤔 The plugin itself should have `tree` as an `mdast.Root` node, everything inside the plugin should be able to infer the types from there (with minimal to no additional annotations). Just make sure `remark-directive` is in the type scope so it can add the directive node types into `mdast`.

Reference resources:
1. https://unifiedjs.com/learn/guide/syntax-trees-typescript
2. https://unifiedjs.com/learn/recipe/tree-traversal-typescript
3. https://unifiedjs.com/learn/recipe/narrow-node-typescript
4. https://unifiedjs.com/learn/recipe/build-a-syntax-tree/

**Reply by Firgrep** [reply]
Yep, I try to import the type root from (I've tried each of these in turn):

```ts
import { type Root } from '@types/mdast';
import type {Root} from 'mdast' // which is what I see used here: https://unifiedjs.com/learn/recipe/tree-traversal-typescript/
import type {Root} from 'remark-directive/lib' 
import type {Root} from 'mdast-util-definitions/lib'
import type {Node as MDASTNode } from 'mdast-util-definitions/lib'
type Root = import('mdast').Root

function adminitionPlugin() {
    return (tree: Root) => {
        visit(tree, (node: ContainerDirective | LeafDirective | TextDirective ) => {
          ....etc...
```

And they all bug out: `Type instantiation is excessively deep and possibly infinite.` 🤔

**Reply by ChristianMurphy** [reply]
What version of `mdx`, `@types/mdast` and `remark-directive` are you using?
The unified ecosystem is the midst of a major version update, which includes new stricter typings, mixing versions of the typings causes a variety of issues (like: https://github.com/unifiedjs/unified/issues/227) potentially including the issue you describe.

A major version of `mdx` is coming, when it is released: `mdx` version 3, `@types/mdast` version 4, and `remark-directive` version 3 can be used together.

Right now: `mdx` version 2, `@types/mdast` version 3, and `remark-directive` version 2 are type compatible.
Hold on these versions until the next MDX release.

**Reply by Firgrep** [reply]
```
// from app's package.json
"@mdx-js/mdx": "^2.3.0",
"remark-directive": "^3.0.0",
"@types/mdx": "^2.0.7",

// presumably imported by a dependency, I find in node_modules/@types/mdast/package.json
"name": "@types/mdast",
"version": "3.0.12",
```

I switched `remark-directive` from 3 to 2, but TS still warns the same thing. 

I had a look at the `mdx-editor` I'm using, checking its [dependencies](https://github.com/mdx-editor/editor/blob/main/package.json), and they seem fine. 

Ok, thanks for the heads up about the incoming new versions! I may just temporarily ts-ignore it then until `mdx` (v3), `@types/mdast` (v4) and `remark-directive` (v3) are out.

**Reply by ChristianMurphy** [reply]
Another idea would be upgrade your TypeScript version.
TypeScript had a bug a while back that would trigger a stall/infinite types when used with `unist-util-visit` (ref: https://github.com/syntax-tree/unist-util-visit/issues/33 and https://github.com/microsoft/TypeScript/issues/46900)

If the issue persists after upgrading TypeScript consider sharing a sandbox with an example so others can replicate the issue.

**Reply by Firgrep** [reply]
Thanks for your suggestions! I had TS 5.1.6 and I've updated to the latest 5.2.2 (both in the app and globally), but the stall/infinite bug persists. 

I set up a stackblitz and, surprisingly, I'm not getting any stall/infinity errors there with `import type {Root} from 'mdast'`. I then added all the packages from my app into the stackblitz sample and still no error... 

But on my end locally, the bug's still there. So I've tried the following:
- I then deleted my local `node_modules` and re-installed everything.
- I updated ts globally to latest. 
- I updated `node.js` to 18.18 to match the version on stackblitz. 
- `tsconfig.json` on the stackblitz was nearly identical to my local one except `"target": "ES2017"` vs `"es5"`. I switched that over but the bug bubbled up all the same eventually.

Not sure where to go from here. Any ideas?

**Reply by ChristianMurphy** [reply]
🤔 without a way to replicate it, it is a bit tricky.

Some assorted ideas worth trying:
1. check your _package-lock.json_ file locally, search for `@types/mdast` to see if conflicting versions have somehow been introduced
2. Try enabling some of TypeScript's stricter options like https://github.com/mdx-js/mdx/blob/a2c86936a99e695b5175f893d481171c8ce572fb/tsconfig.json#L12 and https://github.com/mdx-js/mdx/blob/a2c86936a99e695b5175f893d481171c8ce572fb/tsconfig.json#L6-L7 sometimes stronger config options can narrow the possible type down.
3. > I updated ts globally to latest
   
   try adding the latest typescript to your _local_ project folder, it's possible there is a local version which is older getting picked up by the npm scripts.

**Reply by Firgrep** [reply]
Thanks again for your help! 

1. Checking the `package-lock.json`, I've found two different versions of `@types/mdast`. Everything uses ^3.0.0 except `remark-directive` which installs ^4.0.0. But I already tried rolling `remark-directive` back to version 2 and that didn't help. Checking the stackblitz example, I see that it uses a later version of version 3 (version `3.0.13`) than what I have (version `3.0.12`). There are also twice as many hits in the `package-lock.json` in comparison to mine. I tried running `npm install` to see if that updated the `@types/mdast` but it didn't. I read that it's a bad idea to delete the `package-lock.json` file. What do you reckon?

remark-directive's 
```
"node_modules/remark-directive/node_modules/@types/mdast": {
      "version": "4.0.1",
      "resolved": "https://registry.npmjs.org/@types/mdast/-/mdast-4.0.1.tgz",
      "integrity": "sha512-IlKct1rUTJ1T81d8OHzyop15kGv9A/ff7Gz7IJgrk6jDb4Udw77pCJ+vq8oxZf4Ghpm+616+i1s/LNg/Vh7d+g==",
      "dependencies": {
        "@types/unist": "*"
      }
    },
```

And the rest use this
```
"node_modules/@types/mdast": {
      "version": "3.0.12",
      "resolved": "https://registry.npmjs.org/@types/mdast/-/mdast-3.0.12.tgz",
      "integrity": "sha512-DT+iNIRNX884cx0/Q1ja7NyUPpZuv0KPyL5rGNxm1WC1OtHstl7n4Jb7nk+xacNShQMbczJjt8uFzznpp6kYBg==",
      "dependencies": {
        "@types/unist": "^2"
      }
    },
```

2 & 3. I had `strict` already enabled, but I also added the two others you mentioned. Local `typescript` was also updated to the latest. Bug still persists :/

**Reply by ChristianMurphy** [reply]
> I read that it's a bad idea to delete the package-lock.json file. What do you reckon?

It can be a good idea, but have a backup.
Make sure the current _package-lock.json_ is committed in source control, just in case there is an issue.
Then clear both _package-lock.json_ and _node_modules_ to do a clean reinstall.

> But I already tried rolling remark-directive back to version 2 and that didn't help. 

`remark-directive` version 2, depends on `@types/mdast` version 3 https://github.com/remarkjs/remark-directive/blob/52501b7de6d700f6b8263aeb6f1ed8086242e437/package.json#L36
It is possible the install didn't run, or you may have another package which depends on `@types/mdast` version 4.

> There are also twice as many hits in the package-lock.json in comparison to mine. 

I don't understand this comment, what is a hit?
Do you mean references to `@types/mdast` in the lockfile?
Are you using npm's lockfile version 1, 2, or 3?

**Reply by Firgrep** [reply]
Okay, so I was writing my report (see details below) after I had done a fresh re-install when I just happened to remove some things and the infinity bug disappeared. I'm trying to re-create it and it comes back when the types for `node` here are put in:

```ts
return (tree: Root) => {
        visit(tree, (node: ContainerDirective | LeafDirective | TextDirective) => {
```

If I remove the `node` types, the error vanishes.

```ts
return (tree: Root) => {
        visit(tree, (node) => {
```

The thing is, I could've swore I tried this before! If I add those types to the node in the [stackblitz](https://stackblitz.com/edit/stackblitz-starters-dy4ckd?file=app%2Ftype-test.ts), it throws a different error (types are incompatible instead of infinity bug). 

I `git reset --hard HEAD` to go back to my old `package-lock.json`, delete `node_modules` and re-install it and attempt the same thing. No bug! 

A few things I can think of is that I might have accidentally left a `// @ts-ignore` line above the `return (tree: Root) => {` line when I was trying different things, or that the typescript update had changed things along the way.

Puuhhhhhh! 😅 Well, thanks again for your incredible patience and help, @ChristianMurphy 


<details><summary>previously...</summary>
> Then clear both package-lock.json and node_modules to do a clean reinstall.

Okay, tried this. No avail. :/

> It is possible the install didn't run, or you may have another package which depends on @types/mdast version 4.

Since the stackblitz version is not throwing up any errors, I'm assuming everything is fine in terms of dependencies. 

> I don't understand this comment, what is a hit?
Do you mean references to @types/mdast in the lockfile?

Apologies. Yes, I meant the number of times `ctrl + f` found entries of `@types/mdast`. 

> Are you using npm's lockfile version 1, 2, or 3?

Version 3. I see that the stackblitz one is using version 2. I see it's using `npm 9.4.2`, whereas I have `npm 10.1.0` on my system. By the way, here's the [stackblitz with all packages from my app](https://stackblitz.com/edit/stackblitz-starters-dy4ckd?file=app%2Ftype-test.ts).
</details>

**Reply by wooorm** [reply]
The types of `visit` before could be slow.
This was solved in latest: https://github.com/syntax-tree/unist-util-visit-parents/commit/766107b5f4b7dd8cebf0aaf10d5da620124d170f.

```
return (tree: Root) => {
        visit(tree, (node: ContainerDirective | LeafDirective | TextDirective) => {
```

This is typed wrong. `node` can be any descendant of `Root`, not only these directives.


```
return (tree: Root) => {
        visit(tree, (node) => {
```

This is typed correctly. `visit` is smart enough to know the type of `node`. You should not specify it.
